ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -65,76 +46,80 @@ let call_api_function_with_alert ~__context ~msg ~cls ~obj_uuid ~body
           raise err
       )
 
-let create_as_necessary ~__context ~host =
-  match sync_required ~__context ~host with
-  | Some cluster_ref -> (* assume pool autojoin set *)
-    let network = get_network_internal ~__context ~self:cluster_ref in
-    let (pifref,pifrec) = Xapi_clustering.pif_of_host ~__context network host in
-    fix_pif_prerequisites ~__context (pifref,pifrec);
-
-    (* Best effort Cluster_host.create *)
-    let body = Printf.sprintf "Unable to create cluster host on %s."
-        (Db.Host.get_name_label ~__context ~self:host) in
-    let obj_uuid = Db.Host.get_uuid ~__context ~self:host in
-    call_api_function_with_alert ~__context
-      ~msg:Api_messages.cluster_host_creation_failed
-      ~cls:`Host ~obj_uuid ~body
-      ~api_func:(fun rpc session_id -> ignore
-          (Client.Client.Cluster_host.create rpc session_id cluster_ref host pifref))
-  | None -> ()
-
-let resync_host ~__context ~host =
-  create_as_necessary ~__context ~host;
-  match (find_cluster_host ~__context ~host) with
-  | None              -> () (* no clusters exist *)
-  | Some cluster_host ->    (* cluster_host and cluster exist *)
-    (* Cluster_host.enable unconditionally invokes the low-level enable operations and is idempotent. *)
-    if Db.Cluster_host.get_enabled ~__context ~self:cluster_host then begin
-      (* RPU reformats partition, losing service status, never re-enables clusterd *)
-      debug "Cluster_host %s is enabled, starting up xapi-clusterd" (Ref.string_of cluster_host);
-      Xapi_clustering.Daemon.enable ~__context;
-
-      (* Best-effort Cluster_host.enable *)
-      let body = Printf.sprintf "Unable to create cluster host on %s."
-          (Db.Host.get_name_label ~__context ~self:host) in
-      let obj_uuid = Db.Host.get_uuid ~__context ~self:host in
-      call_api_function_with_alert ~__context
-        ~msg:Api_messages.cluster_host_enable_failed
-        ~cls:`Host ~obj_uuid ~body
-        ~api_func:(fun rpc session_id -> ignore
-            (Client.Client.Cluster_host.enable rpc session_id cluster_host))
-    end
-
-let create ~__context ~cluster ~host ~pif =
-  (* TODO: take network lock *)
+(* Create xapi db object for cluster_host, resync_host calls clusterd *)
+let create_internal ~__context ~cluster ~host ~pIF : API.ref_Cluster_host =
   with_clustering_lock (fun () ->
       assert_operation_host_target_is_localhost ~__context ~host;
       assert_cluster_host_can_be_created ~__context ~host;
-      assert_pif_attached_to ~host ~pif ~__context;
+      assert_pif_attached_to ~host ~pIF ~__context;
```

Comment:
this will fail, should be part of join.

Reply:
Can you add a unit test that tries to create a cluster host object when the PIF exists, but the IP is empty?

Reply:
We already test this in `test_clustering.test_assert_pif_prerequisites`, ensuring `pif_has_no_network_configuration` is raised if no IP exists.

Reply:
We should have a unit test that tries to call Xapi_cluster_host.create and succeed with an empty IP.

Reply:
What should the mock RPC actually do? If it just returns a success, the code moves onto PIF prerequisites, at which point it raises `Api_errors.pif_has_no_network_configuration`, and does the same as the test mentioned above.

Reply:
Lets create a test for `Xapi_cluster_host.create_as_necessary` then, which should succeed even if you have no network.

Reply:
And then you can assign some IP to the PIF, and call resync_host, and check that it works in a second step in the unit test